scipy/stats/_survival.py

```diff
@@ -196,3 +189,60 @@ def _ecdf_right_censored(sample):
     sf = np.cumprod((n - d) / n)
     cdf = 1 - sf
     return ECDFResult(t, cdf, sf)
+
+
+def _kaplan_meier(sample):
+    tod = sample._uncensored  # time of "death"
+    tol = sample._right  # time of "loss"
+    times = np.concatenate((tod, tol))
+    died = np.asarray([1]*tod.size + [0]*tol.size)
+
+    # sort by times
+    i = np.argsort(times)
+    times = times[i]
+    died = died[i]
+    at_risk = np.arange(times.size, 0, -1)
+    return times, died, at_risk
+
```

Comment:
I got your PR in and rebased locally. I am working on it. I get the part 1 and 3 and now playing to get 2 nicely. Thanks for the pointers this is helpful 😃

Reply:
For reference this is what I have at the moment.

Step 0.

```python
sample = stats.CensoredData.right_censored(t2, np.logical_not(d2))
control = stats.CensoredData.right_censored(t3, np.logical_not(d3))
sample_tot = stats.CensoredData.right_censored(t2+t3, np.logical_not(d2+d3))
```

Step 1.

```python
n_died_sample = sample._uncensored.size
n_died_control = control._uncensored.size
```


Step 3.

```python
res = stats.ecdf(sample_tot)
times_sample_tot = res.x[res.sf._d.astype(bool)]
at_risk_sample_tot = res.sf._n[res.sf._d.astype(bool)]
```

Step 2.

First option is to start with the current code in the PR
```python
def _log_rank_tables(sample):
    tod = sample._uncensored  # time of "death"
    tol = sample._right  # time of "loss"
    times = np.concatenate((tod, tol))
    died = np.asarray([1]*tod.size + [0]*tol.size)
    # sort by times
    i = np.argsort(times)
    times = times[i]
    died = died[i]
    at_risk = np.arange(times.size, 0, -1)

    idx_died = died.astype(bool)
    times = times[idx_died]
    at_risk = at_risk[idx_died]
    
    times, dead_per_times = np.unique(sample._uncensored, return_counts=True)
    
    return times, at_risk, dead_per_times


times_sample, died_sample, at_risk_sample = _log_rank_tables(sample)
times_control, died_control, at_risk_control = _log_rank_tables(control)
times_sample_tot, _, at_risk_sample_tot = _log_rank_tables(sample_tot)
```

Now if we don't want this `_log_rank_tables`, I am trying to play with:

```python
tod = sample._uncensored  # time of "death"
tol = sample._right  # time of "loss"
times = np.concatenate((tod, tol))
i = np.argsort(times)
times = times[i]

times_, counts_per_times = np.unique(times, return_counts=True)
at_risk_times_ = len(sample)-np.cumsum(counts_per_times)
at_risk_times_ = np.concatenate([[len(sample)], at_risk_times_])
idx = np.searchsorted(times_, dead_times)
at_risk_sample = at_risk_times_[idx]

...

# do the same for control
```

For both code the following step would be what's in the PR

```python
at_risk_ = at_risk_sample_tot.copy()
idx = np.searchsorted(times_sample_tot, times_sample)
at_risk_[idx] = at_risk_sample_tot[idx] - at_risk_sample
idx = np.searchsorted(times_sample_tot, times_control)
at_risk_[idx] = at_risk_control

at_risk_control = at_risk_
at_risk_sample = at_risk_sample_tot - at_risk_

# and with dead_per_times and some search sorted we can have all Ot
```

But I am trying to get to the final `at_risk` in one go with the above.

Reply:
```python
[np.where(times >= x)[0].size for x in times_sample_tot]
```

Reply:
Interesting, I will play with this.

Otherwise what do you think about the one liner? Or a similar variant. Maybe a bit slower, not even sure and I would need to check, but might be simpler to get.

Reply:
>  (That's why this is not trivial to do with vectorized operations. It would be easy if we wrote a big loop, but I don't want to do that.)

So it's against the rules : )

Reply:
This is the way. (Yes watching the last season)

Reply:
FYI some quick timing for the dummy solution. Not that bad
```python
times = rng.integers(0, 100, size=(10000))
times_sample_tot = rng.integers(0, 100, size=(10000))

%timeit -n 10 [np.where(times >= x)[0].size for x in times_sample_tot]
# 93.1 ms ± 1.14 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

Reply:
The result with `bisect_left` is quite nice! (I tried to do the bisection on on the reduced array, but it did not do anything. I bet the underlying implementation is quite fast to really make a dent here.)

```python
def toto3(times, times_sample_tot):
    at_risk = []
    n_times = len(times)
    for x in times_sample_tot:
        idx_ = bisect.bisect_left(times, x)
        at_risk.append(n_times-idx_)

%timeit -n 10 toto3(times, times_sample_tot)
5.69 ms ± 79.3 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

What do you think? The numerical complexity seem contained here no?

Reply:
This is because `ecdf` will give you the number at risk in the combined sample at each of the unique times. I thought this was supposed to give you the number at risk in one of the groups at those same unique times. (Then you would be able to get the number at risk in the other group by subtracting.) It's producing an array with 10000 elements, but there are only 100 unique ones.

Reply:
I was forgetting that we already have called `ecdf` on the combined sample. This gives you the unique times at which you need to get the number at risk. Using those would make the solution in https://github.com/tupui/scipy/pull/19#discussion_r1143819381 simpler.

And if we want to eliminate:
```python3
x_unique, x_counts = np.unique(x, return_counts=True)
n_risk_x = x.size - np.cumsum(x_counts[:-1])
```
We can just call `stats.ecdf` on the X group.

So if we are going for a good balance between simplicity and speed:

```python3
x = rng.permuted([1, 1, 2, 3, 4, 5, 6, 6, 6])
y = rng.permuted([-1, 0, 0.5, 1.5, 3, 4.5, 5])
z = np.concatenate([x, y])

dist = stats.ecdf(z)
dist_x = stats.ecdf(x)
at_risk_x = np.insert(dist_x.sf._n[:-1], 0, len(x))
i = np.searchsorted(dist_x.x, dist.x)  # or use `interpolate_1d`
n_risk_x_at_all_times = at_risk_x[i]

np.testing.assert_allclose(n_risk_x_at_all_times, f(x, y))
```

I guess I should have led with that.